### PR TITLE
ROS2 migration.

### PR DESCRIPTION
**Pull request by clalancette**
This is an omnibus port of kdl_parser to ROS2.  The original
code for this came from the following commits from https://github.com/ros2/robot_model:

7c17f1e0e20b58de71d2c24a85c0c29128d25e45
bb68744f035873101f4d97abff473d3fc9471a5e
ae9e7dbbe7f64df30c29993956281e9ad96ae905
6bd94a8d9d26bee289b43949278075486a0e23d5
a2d3b1290dff50e516a510a2e1e463f987462e79
cdcf138ad41f2f99ccf781de8999699f8ad94d45
762cad7eb2b05c9982665f2c284cccdd77703971
7a194ed90f6c1df66155a07e450224d6a07f1802
558ecf2412be6f2f3d4b25f435748c5a3771a9a8
c177e3921e592d2b31d0e7edc23b3b4748a53e1a

connects to ros2/ros2#419

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

### CODE REVIEW

File: kdl_parser/CMakeLists.txt [hunk 1]
```diff
@@ -1,51 +1,69 @@
...
 
-catkin_package(
-  LIBRARIES ${PROJECT_NAME} ${orocos_kdl_LIBRARIES}
-  INCLUDE_DIRS include
-  CATKIN_DEPENDS roscpp rosconsole urdf
-  DEPENDS orocos_kdl TinyXML
-)
+add_library(
+  ${PROJECT_NAME}
+  SHARED
```
Comment (mikaelarguedas) [inline kdl_parser/CMakeLists.txt]: Can use ament_cmake_ros here as well

File: kdl_parser/CMakeLists.txt [hunk 2]
```diff
@@ -1,51 +1,69 @@
...
-  target_link_libraries(test_inertia_rpy ${PROJECT_NAME})
+if(BUILD_TESTING)
+  find_package(ament_cmake_cppcheck REQUIRED)
+  find_package(ament_cmake_cpplint REQUIRED)
+  find_package(ament_cmake_lint_cmake REQUIRED)
+  find_package(ament_cmake_uncrustify REQUIRED)
+  ament_cppcheck()
+  ament_cpplint()
+  ament_lint_cmake()
+  ament_uncrustify()
```
Comment (mikaelarguedas) [inline kdl_parser/CMakeLists.txt]: use ament_lint_auto

File: kdl_parser/CMakeLists.txt [hunk 3]
```diff
@@ -1,51 +1,69 @@
...
 endif()
 
-# How does CATKIN do this?
-#rosbuild_add_rostest(${PROJECT_SOURCE_DIR}/test/test_kdl_parser.launch)
-install(TARGETS ${PROJECT_NAME} 
-  DESTINATION ${CATKIN_PACKAGE_LIB_DESTINATION})
-
-install(DIRECTORY include/${PROJECT_NAME}/
-  DESTINATION ${CATKIN_PACKAGE_INCLUDE_DESTINATION})
+ament_export_dependencies(orocos_kdl)
```
Comment (mikaelarguedas) [inline kdl_parser/CMakeLists.txt:62]: this should export urdf as well as it's using urdf types in it's public API right?

Comment (mikaelarguedas) [reply kdl_parser/CMakeLists.txt:62]: these dependencies should also be exported in the package.xml (or at least urdfdom_headers in the case of urdf)

File: kdl_parser/test/test_inertia_rpy.cpp [hunk 4]
```diff
@@ -1,40 +1,70 @@
+/*********************************************************************
+* Software License Agreement (BSD License)
+*
+*  Copyright (c) 2008, Willow Garage, Inc.
```
Comment (mikaelarguedas) [inline kdl_parser/test/test_inertia_rpy.cpp]: This should be Open Source Robotics Foundation 2015 (https://github.com/ros/kdl_parser/commits/kinetic-devel/kdl_parser/test/test_inertia_rpy.cpp)

**Review by mikaelarguedas — COMMENTED** [review]
A few minor comments to address otherwise lgtm

File: kdl_parser/CMakeLists.txt (hunk 1 above)
Comment (clalancette) [reply kdl_parser/CMakeLists.txt]: Yep, done in 4c440e8

File: kdl_parser/CMakeLists.txt (hunk 2 above)
Comment (clalancette) [reply kdl_parser/CMakeLists.txt]: Done in 4c440e8

File: kdl_parser/CMakeLists.txt (hunk 3 above)
Comment (clalancette) [reply kdl_parser/CMakeLists.txt:62]: From looking at it, I think `orocos_kdl` and `urdfdom_headers` are the exported dependencies, while `urdf` is a build dependency.  I've marked them all as such in CMakeLists.txt and package.xml in 4c440e8

File: kdl_parser/test/test_inertia_rpy.cpp (hunk 4 above)
Comment (clalancette) [reply kdl_parser/test/test_inertia_rpy.cpp]: Yeah, it looks like Jackie wrote that file initially (with no copyright notice), and then @Karsten1987 added the copyright notice earlier this year.  So I agree with you there, and I've changed it.  It still doesn't pass copyright checks, so I disabled it in 4c440e8

File: kdl_parser/CMakeLists.txt (hunk 3 above)
Comment (mikaelarguedas) [reply kdl_parser/CMakeLists.txt:62]: >  I think orocos_kdl and urdfdom_headers are the exported dependencies

Yes that matches what I guessed when reviewing :+1:

File: kdl_parser/package.xml [hunk 5]
```diff
@@ -36,6 +36,9 @@
   <test_depend>ament_lint_auto</test_depend>
   <test_depend>ament_lint_common</test_depend>
 
+  <build_export_depend version_gte="1.3.0">orocos_kdl</build_export_depend>
```
Comment (mikaelarguedas) [inline kdl_parser/package.xml]: Nit: using a single `depend` tag would allow you to not have to duplicate this with the version number in 3 places

**Review by mikaelarguedas — DISMISSED** [review]
lgtm

**Comment by mikaelarguedas** [comment]
same remark as https://github.com/ros2/urdf/pull/1#issuecomment-347701962 about preserving the history of the changes

**Comment by clalancette** [comment]
OK, rebased to split it up as we discussed.  CI is in ros2/ros2#419.  Ready for another review.

File: kdl_parser/src/check_kdl_parser.cpp [hunk 6]
```diff
@@ -34,45 +34,47 @@
...
 /* Author: Wim Meeussen */
 
-#include "kdl_parser/kdl_parser.hpp"
-#include <kdl/chainfksolverpos_recursive.hpp>
-#include <kdl/frames_io.hpp>
-#include <urdf/model.h>
 #include <iostream>
+#include <string>
+
+#include "urdf/model.h"
```
Comment (mikaelarguedas) [inline kdl_parser/src/check_kdl_parser.cpp]: nit: alphabetical

File: kdl_parser/src/kdl_parser.cpp [hunk 7]
```diff
@@ -35,132 +35,138 @@
...
 {
-  Frame F_parent_jnt = toKdl(jnt->parent_to_joint_origin_transform);
-
-  switch (jnt->type){
-  case urdf::Joint::FIXED:{
-    return Joint(jnt->name, Joint::None);
+  KDL::Frame F_parent_jnt = toKdl(jnt->parent_to_joint_origin_transform);
+
+  switch (jnt->type) {
+    case urdf::Joint::FIXED: {
```
Comment (mikaelarguedas) [inline kdl_parser/src/kdl_parser.cpp]: brackets in case are redundant and result in weird indent, please remove the square brackets and use 2-space indent

File: kdl_parser/src/kdl_parser.cpp [hunk 8]
```diff
@@ -35,132 +35,138 @@
...
 
 
 // recursive function to walk through tree
-bool addChildrenToTree(urdf::LinkConstSharedPtr root, Tree& tree)
+bool addChildrenToTree(urdf::LinkConstSharedPtr root, KDL::Tree & tree)
 {
-  std::vector<urdf::LinkSharedPtr > children = root->child_links;
-  ROS_DEBUG("Link %s had %i children", root->name.c_str(), (int)children.size());
+  std::vector<urdf::LinkSharedPtr> children = root->child_links;
+  fprintf(stderr, "Link %s had %i children\n",
```
Comment (mikaelarguedas) [inline kdl_parser/src/kdl_parser.cpp]: this should not go upsteam and thus should be in another commit

File: kdl_parser/test/test_inertia_rpy.cpp [hunk 9]
```diff
@@ -1,5 +1,42 @@
+/*********************************************************************
+* Software License Agreement (BSD License)
+*
+*  Copyright (c) 2015 Open Source Robotics Foundation
```
Comment (mikaelarguedas) [inline kdl_parser/test/test_inertia_rpy.cpp]: please use "Open Source Robotics Foundation Inc." like the rest of the codebase

File: kdl_parser/test/test_inertia_rpy.cpp [hunk 10]
```diff
@@ -1,5 +1,42 @@
...
+*  modification, are permitted provided that the following conditions
+*  are met:
+*
+*   * Redistributions of source code must retain the above copyright
+*     notice, this list of conditions and the following disclaimer.
+*   * Redistributions in binary form must reproduce the above
+*     copyright notice, this list of conditions and the following
+*     disclaimer in the documentation and/or other materials provided
+*     with the distribution.
+*   * Neither the name of the Willow Garage nor the names of its
```
Comment (mikaelarguedas) [inline kdl_parser/test/test_inertia_rpy.cpp]: this should be "Open Source Robotics Foundation"

File: kdl_parser/test/test_kdl_parser.cpp [hunk 11]
```diff
@@ -55,36 +52,32 @@ class TestParser : public testing::Test
...
   }
 
-  ASSERT_TRUE(treeFromFile(g_argv[g_argc-1], my_tree));
+  ASSERT_TRUE(kdl_parser::treeFromFile(g_argv[g_argc - 1], my_tree));
   ASSERT_EQ(my_tree.getNrOfJoints(), 8);
   ASSERT_EQ(my_tree.getNrOfSegments(), 16);
   ASSERT_TRUE(my_tree.getSegment("dummy_link") == my_tree.getRootSegment());
   ASSERT_EQ(my_tree.getRootSegment()->second.children.size(), (unsigned int)1);
   ASSERT_TRUE(my_tree.getSegment("base_link")->second.parent == my_tree.getRootSegment());
   ASSERT_EQ(my_tree.getSegment("base_link")->second.segment.getInertia().getMass(), 10.0);
```
Comment (mikaelarguedas) [inline kdl_parser/test/test_kdl_parser.cpp]: all these should have the liltteral value on the expected (left) side not the actual (right) side

File: kdl_parser/package.xml [hunk 12]
```diff
@@ -20,18 +20,26 @@
...
 
   <build_depend version_gte="1.3.0">orocos_kdl</build_depend>
-  <build_depend>rosconsole</build_depend>
-  <build_depend>roscpp</build_depend>
   <build_depend>urdf</build_depend>
-  <build_depend>cmake_modules</build_depend>
-  <build_depend>rostest</build_depend>
+  <build_depend>urdfdom_headers</build_depend>
+  <build_depend>tinyxml</build_depend>
+  <build_depend>tinyxml_vendor</build_depend>
```
Comment (mikaelarguedas) [inline kdl_parser/package.xml]: dependencies should be listed in alphabetical order

File: kdl_parser/src/kdl_parser.cpp [hunk 13]
```diff
@@ -39,10 +39,9 @@
 #include <string>
 #include <vector>
 
-#include <urdf/model.h>
-#include <urdf/urdfdom_compatibility.h>
-#include <kdl/frames_io.hpp>
-#include <ros/console.h>
+#include "urdf/model.h"
+#include "urdf/urdfdom_compatibility.h"
+#include "kdl/frames_io.hpp"
```
Comment (mikaelarguedas) [inline kdl_parser/src/kdl_parser.cpp]: include order

**Review by mikaelarguedas — COMMENTED** [review]
a few things belong in the wrong commits or should be fixed in the copyrights. The rest is nitpicks

**Comment by clalancette** [comment]
I think I've fixed all of the nits pointed out here now.

File: kdl_parser/include/kdl_parser/visibility_control.hpp [hunk 14]
```diff
@@ -0,0 +1,56 @@
+// Copyright 2015 Open Source Robotics Foundation, Inc.
```
Comment (Karsten1987) [inline kdl_parser/include/kdl_parser/visibility_control.hpp]: year

Comment (clalancette) [reply kdl_parser/include/kdl_parser/visibility_control.hpp]: Actually, for that matter, this should be the BSD license to match the rest of the code anyway.  I'll fix both of those.

**Comment by mikaelarguedas** [comment]
@clalancette github has a weird display but I made my review on a per commit basis so you need to browse commits individually to see all review comments. Several of tham have not been addressed yet

File: kdl_parser/src/kdl_parser.cpp [hunk 15]
```diff
@@ -170,42 +176,50 @@ bool treeFromParam(const string& param, Tree& tree)
...
+  }
 
-  tree = Tree(robot_model.getRoot()->name);
+  tree = KDL::Tree(robot_model.getRoot()->name);
 
   // warn if root link has inertia. KDL does not support this
-  if (robot_model.getRoot()->inertial)
-    ROS_WARN("The root link %s has an inertia specified in the URDF, but KDL does not support a root link with an inertia.  As a workaround, you can add an extra dummy link to your URDF.", robot_model.getRoot()->name.c_str());
+  if (robot_model.getRoot()->inertial) {
+    auto error_msg = std::string("The root link") +
```
Comment (mikaelarguedas) [inline kdl_parser/src/kdl_parser.cpp]: this is C++11 and cannot go upstream as is

File: kdl_parser/CMakeLists.txt [hunk 16]
```diff
@@ -1,51 +1,66 @@
...
+  ${orocos_kdl_LIBRARIES}
+  ${urdf_LIBRARIES})
 
-add_library(${PROJECT_NAME} src/kdl_parser.cpp)
-target_link_libraries(${PROJECT_NAME}
-  ${TinyXML_LIBRARIES} ${orocos_kdl_LIBRARIES} ${catkin_LIBRARIES}
-)
+if(WIN32)
+  target_compile_definitions(${PROJECT_NAME} PRIVATE "KDL_PARSER_BUILDING_DLL")
+endif()
```
Comment (mikaelarguedas) [inline kdl_parser/CMakeLists.txt:42]: This should go in the windows compatibility commit right ?

File: kdl_parser/src/kdl_parser.cpp (hunk 8 above)
Comment (mikaelarguedas) [reply kdl_parser/src/kdl_parser.cpp]: (not to self) this is now resolved

File: kdl_parser/src/kdl_parser.cpp [hunk 17]
```diff
@@ -35,132 +35,138 @@
...
 
 // recursive function to walk through tree
-bool addChildrenToTree(urdf::LinkConstSharedPtr root, Tree& tree)
+bool addChildrenToTree(urdf::LinkConstSharedPtr root, KDL::Tree & tree)
 {
-  std::vector<urdf::LinkSharedPtr > children = root->child_links;
-  ROS_DEBUG("Link %s had %i children", root->name.c_str(), (int)children.size());
+  std::vector<urdf::LinkSharedPtr> children = root->child_links;
+  ROS_DEBUG("Link %s had %i children",
+    root->name.c_str(), static_cast<int>(children.size()));
```
Comment (mikaelarguedas) [inline kdl_parser/src/kdl_parser.cpp]: could this just be printed as a `size_t` removing the need for an explicit cast?

File: kdl_parser/src/check_kdl_parser.cpp (hunk 6 above)
Comment (clalancette) [reply kdl_parser/src/check_kdl_parser.cpp]: Should be fixed in this commit now.

File: kdl_parser/src/kdl_parser.cpp (hunk 7 above)
Comment (clalancette) [reply kdl_parser/src/kdl_parser.cpp]: In this particular case, the brackets are being used to keep the scope of the `axis` variable small.  While there are other ways to do this, I'd prefer not to change this right now.

File: kdl_parser/src/kdl_parser.cpp (hunk 17 above)
Comment (clalancette) [reply kdl_parser/src/kdl_parser.cpp]: Yeah, I've changed this now.

File: kdl_parser/src/kdl_parser.cpp (hunk 15 above)
Comment (clalancette) [reply kdl_parser/src/kdl_parser.cpp]: Yeah, good point, I fixed this.

File: kdl_parser/CMakeLists.txt (hunk 16 above)
Comment (clalancette) [reply kdl_parser/CMakeLists.txt:42]: Yeah, moved into that commit now.

File: kdl_parser/package.xml (hunk 12 above)
Comment (clalancette) [reply kdl_parser/package.xml]: Fixed now.

File: kdl_parser/package.xml (hunk 5 above)
Comment (clalancette) [reply kdl_parser/package.xml]: Switched to `<depend>`

**Comment by clalancette** [comment]
> @clalancette github has a weird display but I made my review on a per commit basis so you need to browse commits individually to see all review comments. Several of tham have not been addressed yet

Thanks, I did indeed miss that before.  I think I've now searched them all out and fixed them all.

File: kdl_parser/src/kdl_parser.cpp (hunk 7 above)
Comment (mikaelarguedas) [reply kdl_parser/src/kdl_parser.cpp]: sounds good :+1: thanks for clarifying

File: kdl_parser/include/kdl_parser/visibility_control.hpp [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+*  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+*  POSSIBILITY OF SUCH DAMAGE.
+*********************************************************************/
+
+/* This header must be included by all rclcpp headers which declare symbols
+ * which are defined in the rclcpp library. When not building the rclcpp
+ * library, i.e. when using the headers in other package's code, the contents
+ * of this header change the visibility of certain symbols which the rclcpp
+ * library cannot have, but the consuming code must have inorder to link.
+ */
```
Comment (mikaelarguedas) [inline kdl_parser/include/kdl_parser/visibility_control.hpp]: copy-n-paste?

File: kdl_parser/CMakeLists.txt [hunk 19]
```diff
@@ -1,55 +1,66 @@
...
-  target_link_libraries(test_kdl_parser ${PROJECT_NAME})
+install(
+  DIRECTORY include/${PROJECT_NAME}/
+  DESTINATION include/${PROJECT_NAME})
 
-  add_rostest_gtest(test_inertia_rpy test/test_inertia_rpy.launch test/test_inertia_rpy.cpp )
-  target_link_libraries(test_inertia_rpy ${PROJECT_NAME})
+if(BUILD_TESTING)
+  find_package(ament_lint_auto REQUIRED)
+  # FIXME(clalancette): we have to disable copyright checking until
```
Comment (mikaelarguedas) [inline kdl_parser/CMakeLists.txt]: We generally use only "TODO" in the codebase as this is the only one checked by linters so I'd suggest switching this to a TODO

File: kdl_parser/include/kdl_parser/visibility_control.hpp (hunk 18 above)
Comment (clalancette) [reply kdl_parser/include/kdl_parser/visibility_control.hpp]: Yeah, that was from the original port.  I fixed it now.

File: kdl_parser/CMakeLists.txt (hunk 19 above)
Comment (clalancette) [reply kdl_parser/CMakeLists.txt]: Switched to TODO.

File: kdl_parser/test/test_inertia_rpy.cpp [hunk 20]
```diff
@@ -37,14 +37,12 @@
...
-#include <kdl/jntarray.hpp>
-#include <kdl/chainidsolver_recursive_newton_euler.hpp>
-
-#include <ros/ros.h>
-#include <ros/console.h>
 #include "kdl_parser/kdl_parser.hpp"
 
+#include "gtest/gtest.h"
+#include "kdl/jntarray.hpp"
+#include "kdl/chainidsolver_recursive_newton_euler.hpp"
```
Comment (mikaelarguedas) [inline kdl_parser/test/test_inertia_rpy.cpp]: this looks like a style change but given that I consider this more a weird behavior of our config of cpplint rather than an actual Google Styleguide violation I think it's fine to not upstream.
Nit: out of alphabetical order

**Review by mikaelarguedas — APPROVED** [review]
:tada: thanks for iterating so many times with me.
Hopefully this will make the ROS 1 backport much easier

**Comment by clalancette** [comment]
Thanks, merging.

**Comment by clalancette** [comment]
D'oh, I made a mistake here and did squash-n-merge.  I'm going to force push to undo that (nobody can have pulled it), then open a new PR and just 'rebase-n-merge' instead.
